### PR TITLE
Processing vector geometry improvements

### LINKED ISSUE #2563: [FEATURE][processing] Native wedge buffer algorithm
Original commit: https://github.com/qgis/QGIS/commit/b884eb4a71f433eb0afe7a601b80a5ca17d20ed9 by nyalldawson

This algorithm creates wedge shaped buffers from input
points.

The azimuth parameter gives the angle (in degrees) for the
middle of the wedge to point. The buffer width (in degrees)
is specified by the width parameter. Note that the
wedge will extend to half of the angular width either side
of the azimuth direction.

The outer radius of the buffer is specified via outer radius,
and optionally an inner radius can also be specified.

Supports dynamic parameters for azimuth, width, and radius.

The native output from this algorithm is CurvePolygon
geometries, but these may be automatically segmentized to
Polygons depending on the output format.

### LINKED ISSUE #2591: [FEATURE] Variable width buffers
Original commit: https://github.com/qgis/QGIS/commit/9ffae600f4c4473a25083cd354c07cad7fd08ea4 by nyalldawson

Adds geometry methods to create variable width buffers, including
tapered buffers (with a specified start and end diameter) and
variable width buffers from line string m values.

Also adds processing algorithms which expose these methods
to processing.

### PR DESCRIPTION
**Pull request by ghtmtt**
fix #2563 fix #2591 and other small changes

some pictures added, actually many pictures.. I think it is way better to have clear images and less description (a good balance of both of them). Users can get the idea faster IMHO

### CODE REVIEW

File: source/docs/user_manual/processing_algs/qgis/vectorgeometry.rst [hunk 1]
```diff
@@ -518,6 +519,77 @@ Output
...
+``Input layer`` [vector: any]
+  Input vector layer
+
+``Azimuth (degrees from North)`` [number]
+  Angle (in degrees) for the middle of the wedge to point. It can be a unique value
+  (same value for all the features) or it can be taken from features data (different
+  value depending on the feature attribute).
+
+``Wedge width (in degrees)`` [number]
+  Width (in degrees) of the buffer. It can be a unique value (same value for all
```
Comment (DelazJ) [inline source/docs/user_manual/processing_algs/qgis/vectorgeometry.rst]: `Angle (in degrees) of the wedge shaped buffer`

File: source/docs/user_manual/processing_algs/qgis/vectorgeometry.rst [hunk 2]
```diff
@@ -2228,6 +2300,73 @@ Outputs
...
+
+``Segments`` [number]
+  Number of the buffer segments. It can be a unique value (same value for all the
+  features) or it can be taken from features data (different value depending
+  on the feature attribute).
+
+Output
+......
+
+``Buffered layer`` [vector: polygon]
```
Comment (DelazJ) [inline source/docs/user_manual/processing_algs/qgis/vectorgeometry.rst]: ~~layer~~

File: source/docs/user_manual/processing_algs/qgis/vectorgeometry.rst [hunk 3]
```diff
@@ -2228,6 +2300,73 @@ Outputs
...
+
+``Swapped`` [vector: any]
+  Output swapped vector layer
+
+
+.. _qgistaperedbuffer:
+
+Tapered buffers |32|
+--------------------
+Creates tapered buffers along line geometries, using a specified start and end
```
Comment (DelazJ) [inline source/docs/user_manual/processing_algs/qgis/vectorgeometry.rst]: Given that it creates one buffer per feature, better use `tapered buffer` instead.

File: source/docs/user_manual/processing_algs/qgis/vectorgeometry.rst [hunk 4]
```diff
@@ -2228,6 +2300,73 @@ Outputs
...
+   Tapered buffer example
+
+Parameters
+..........
+
+``Input layer`` [vector: line]
+  Input line vector layer.
+
+``Start width`` [number]
+  Starting width of the tapered buffer. It can be a unique value (same value for
```
Comment (DelazJ) [inline source/docs/user_manual/processing_algs/qgis/vectorgeometry.rst]: `Represents the radius of the buffer applied at the start point (or first vertex?) of the line feature.` or something like that

File: source/docs/user_manual/processing_algs/qgis/vectorgeometry.rst [hunk 5]
```diff
@@ -2228,6 +2300,73 @@ Outputs
...
+  all the features) or it can be taken from features data (different value depending
+  on the feature attribute).
+
+``End width`` [number]
+  Ending width of the tapered buffer. It can be a unique value (same value for
```
Comment (DelazJ) [inline source/docs/user_manual/processing_algs/qgis/vectorgeometry.rst]: As above?

File: source/docs/user_manual/processing_algs/qgis/vectorgeometry.rst [hunk 6]
```diff
@@ -518,6 +519,77 @@ Output
...
+Parameters
+..........
+
+``Input layer`` [vector: any]
+  Input vector layer
+
+``Azimuth (degrees from North)`` [number]
+  Angle (in degrees) for the middle of the wedge to point. It can be a unique value
+  (same value for all the features) or it can be taken from features data (different
+  value depending on the feature attribute).
```
[the remark this answers is not in the document]
Comment (DelazJ) [reply source/docs/user_manual/processing_algs/qgis/vectorgeometry.rst]: note: if the DD issue gets agreement, it should be applied to other places so for readability, I ignore the remark in the other reviews. and maybe some changes may be deferred until that agreement...

File: source/docs/user_manual/processing_algs/qgis/vectorgeometry.rst [hunk 7]
```diff
@@ -2228,6 +2300,73 @@ Outputs
...
+
+Output
+......
+
+``Buffered layer`` [vector: polygon]
+  Variable buffer polygon layer
+
+See also
+........
+:ref:`qgisbufferbym`, :ref:`qgisbuffer`
```
Comment (DelazJ) [inline source/docs/user_manual/processing_algs/qgis/vectorgeometry.rst]: I'd also add the wedge one

File: source/docs/user_manual/processing_algs/qgis/vectorgeometry.rst [hunk 8]
```diff
@@ -2317,6 +2456,40 @@ Outputs
...
+
+``Segments`` [number]
+  Number of the buffer segments. It can be a unique value (same value for all the
+  features) or it can be taken from features data (different value depending
+  on the feature attribute).
+
+Output
+......
+
+``Buffered layer`` [vector: polygon]
```
Comment (DelazJ) [inline source/docs/user_manual/processing_algs/qgis/vectorgeometry.rst]: still no layer

File: source/docs/user_manual/processing_algs/qgis/vectorgeometry.rst [hunk 9]
```diff
@@ -2317,6 +2456,40 @@ Outputs
...
+
+Variable width buffer (by m-value) |32|
+---------------------------------------
+Creates variable width buffers along lines, using the m-value of the line geometries
+as the diameter of the buffer at each vertex.
+
+.. figure:: img/variable_buffer_m.png
+   :align: center
+
+   Tapered buffer example
```
Comment (DelazJ) [inline source/docs/user_manual/processing_algs/qgis/vectorgeometry.rst]: tapered?

File: source/docs/user_manual/processing_algs/qgis/vectorgeometry.rst [hunk 10]
```diff
@@ -2228,6 +2300,73 @@ Outputs
...
+``Segments`` [number]
+  Number of the buffer segments. It can be a unique value (same value for all the
+  features) or it can be taken from features data (different value depending
+  on the feature attribute).
+
+Output
+......
+
+``Buffered layer`` [vector: polygon]
+  Variable buffer polygon layer
```
Comment (DelazJ) [inline source/docs/user_manual/processing_algs/qgis/vectorgeometry.rst]: could be nice to have the idea of tapered here instead.

File: source/docs/user_manual/processing_algs/qgis/vectorgeometry.rst [hunk 11]
```diff
@@ -518,6 +519,77 @@ Output
   Final extent of the layer
 
 
+.. _qgiswedgebuffers:
+
+Create wedge buffers |32|
+-------------------------
+Creates wedge shaped buffers from input points.
```
[the remark this answers is not in the document]
Comment (ghtmtt) [reply source/docs/user_manual/processing_algs/qgis/vectorgeometry.rst]: same for me. I'll leave this small description as it is (for now) and maybe enhance it in a future

File: source/docs/user_manual/processing_algs/qgis/vectorgeometry.rst (hunk 6 above)
[the remark this answers is not in the document]
Comment (ghtmtt) [reply source/docs/user_manual/processing_algs/qgis/vectorgeometry.rst]: Ok I made a first layout example with `[number or data defined icon_here]`. Looks nice ;)

File: source/docs/user_manual/processing_algs/qgis/vectorgeometry.rst (hunk 1 above)
Comment (ghtmtt) [reply source/docs/user_manual/processing_algs/qgis/vectorgeometry.rst]: ?

File: source/docs/user_manual/processing_algs/qgis/vectorgeometry.rst [hunk 12]
```diff
@@ -518,6 +519,69 @@ Output
...
+The native output from this algorithm are CurvePolygon geometries, but these may
+be automatically segmentized to Polygons depending on the output format.
+
+Parameters
+..........
+
+``Input layer`` [vector: point]
+  Input point vector layer
+
+``Azimuth (degrees from North)`` [number or data-defined |dataDefined|]
```
[the remark this answers is not in the document]
Comment (ghtmtt) [reply source/docs/user_manual/processing_algs/qgis/vectorgeometry.rst]: mmm I think that the discussion here could become really long :) 

for the moment I removed the `data-defined` *string* and left only the icon.

> whether the option is data-defined or not, the value returned should still be a number to be accepted so [number] is always the right data type.

agreed

> I think that adding the icon should not anyway exempt us from specifying for each concerned parameter description that it's a unique or data-defined value (made a wording in an earlier comment in this PR)

here I'm -0.5 :) just kidding.. but the problem to specify that this parameter can also accept data-defined values is that we can have a lot of repetition in the algorithm description (like in the first commit of this PR). While having just the icon could be too implicit and uses cannot be aware of it.

I don't know how we can solve this...

**Comment by ghtmtt** [comment]
@DelazJ thanks for all the comments. I removed some pictures fro the wedge buffer and replaced by only  2: a general picture (with some weird shape) and another one explaining the azimuth and width parameters

**Comment by ghtmtt** [comment]
@DelazJ Nope **you** where right, see screenshot;) of course I have a lot of layers in the project (point, line, polygon...)

if you give me the OK I will resolve the conflicts and merge this PR. Thanks!

create wedge buffers_106

**Comment by DelazJ** [comment]
Sorry but **YOU WERE RIGHT**. The screenshot you are showing has nothing to do with the 
384f5f9  commit. That commit concerns "Swap X,Y coordinates" alg  that takes all types of geometry and not only point while you are showing wedge buffer screenshot.

**Comment by ghtmtt** [comment]
ahhh.. sorry.. wrong algorithm.. So I was right :tada: ;) description changed.. I'll fix the conflicts and merge the PR. Thanks again @DelazJ

**Pull request merged**